### DISCUSSION TITLE
Dynamically loading GTM on cookie consent

### DISCUSSION THREAD

**Discussion by ehickman-cm** [vercel/next.js#15416] [Help]
Hello, our team is working on a new site that requires running two instances of Google Tag Manager. One GTM container fires on every page load for every visitor, and the other only begins firing if the user accepts cookies.

Our initial approach was to try inserting the tags into the `<Head>` element in the modal component that handles cookie consent, but it seems like there should be a better way to get this to work.

I'm wondering 1), can the [official example for google analytics](https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics) be modified to work with GTM? In trying to implement that example, I kept running into `TypeError: window.gtag is not a function`, even when checking for the window object. (I'm by no means a GTM expert)

And 2) is there a recommended way to dynamically insert script tags? It seems as though you want the scripts to be in the _document file, but since application logic can't happen there, I was wondering what would be considered best practice.

**Comment by rishabhtatia1** [comment]
Hi, did you find any solution?

**Comment by sebastianpikand** [comment]
Hi. I followed:

https://developers.google.com/tag-platform/devguides/privacy
and
https://developers.google.com/tag-platform/devguides/consent

We are using ONLY google tag manager and through it only google analytics. In app.js add :

```
  <Script id="google-tag-manager-consent" strategy="afterInteractive">
      {`
         // Define dataLayer and the gtag function.
         window.dataLayer = window.dataLayer || [];
         function gtag(){dataLayer.push(arguments);}
      
         // Default ad_storage to 'denied' or 'granted'.
         gtag('consent', 'default', {
           'analytics_storage': 'denied'
         });
      `}
    </Script>
```

and then in your cookie consent popup code for "ACCEPT COOKIES" use this code:

```
gtag('consent', 'update', {
      analytics_storage: 'granted',
    });
```

you can toggle the cookie consent style with localStorage setting that keeps track of accepted consent.

After you have added the code load your page and write into console: window.datalayer. You can see data that has been pushed there. If the initial status is denied it will be the first element in datalayer array. GTM keeps pushing events to datalayer, but they are not processed. If user accepts cookie and analytics_storage is "granted", you will start seeing the user in live view as well (use dev env and a different GTM config so you know you are the only user)

**Reply by heijmerikx** [reply]
@sebastianpikand thanks for the post. However, I'm curious. Doesn't this initial analytics_storage setting to 'denied' mean that initial pageviews also aren't tracked? I understand this is probably as intended, but tracking that first hit (without ip address being collected) should be a possibility too, right?

**Comment by marcelgruber** [comment]
@sebastianpikand is there a reason you specifically referenced `strategy="afterInteractive"`? It is the default value.

https://nextjs.org/docs/pages/building-your-application/optimizing/scripts#strategy

**Reply by sebastianpikand** [reply]
Sorry, I don't remember anymore. Overall I have tried all of the strategies and also placing GTM script in different parts of the page, Google Lighthouse test will anyway scream that you have unused js on your site...

**Comment by aleczratiu** [comment]
I'm running into the same issue, each time I refresh the page cookie consent modal is showing again and again. Does anyone faced this issue or should I handle the cookies because is not clear for me..
